crates/oxc_linter/src/rules/promise/no_return_in_finally.rs

```diff
@@ -0,0 +1,118 @@
+use oxc_allocator::Box as OBox;
+use oxc_ast::{
+    ast::{Expression, FunctionBody, Statement},
+    AstKind,
+};
+use oxc_diagnostics::OxcDiagnostic;
+use oxc_macros::declare_oxc_lint;
+use oxc_span::Span;
+
+use crate::{context::LintContext, rule::Rule, utils::is_promise, AstNode};
+
+fn no_return_in_finally_diagnostic(span0: Span) -> OxcDiagnostic {
+    OxcDiagnostic::warn(
+        "eslint-plugin-promise(no-return-in-finally): Disallow return statements in finally",
+    )
+    .with_label(span0)
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct NoReturnInFinally;
+
+declare_oxc_lint!(
+    /// ### What it does
+    ///
+    /// Disallow return statements in finally().
+    ///
+    /// ### Why is this bad?
+    ///
+    /// Disallow return statements inside a callback passed to finally(), since nothing would
+    /// consume what's returned.
+    ///
+    /// ### Example
+    /// ```javascript
+    /// myPromise.finally(function (val) {
+    ///   return val
+    /// })
+    /// ```
+    NoReturnInFinally,
+    correctness,
+);
+
+impl Rule for NoReturnInFinally {
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        let AstKind::CallExpression(call_expr) = node.kind() else {
+            return;
+        };
+
+        if !is_promise(call_expr) {
+            return;
+        }
+
+        let Some(member_expr) = call_expr.callee.get_member_expr() else {
+            return;
+        };
+
+        let Some(prop_name) = member_expr.static_property_name() else {
+            return;
+        };
+
+        if prop_name != "finally" {
+            return;
+        }
+
+        for argument in &call_expr.arguments {
+            let Some(arg_expr) = argument.as_expression() else {
+                continue;
+            };
+            match arg_expr {
+                Expression::ArrowFunctionExpression(arrow_expr) => {
+                    find_return_statement(&arrow_expr.body, ctx);
+                }
+                Expression::FunctionExpression(func_expr) => {
+                    let Some(func_body) = &func_expr.body else {
+                        continue;
+                    };
+                    find_return_statement(func_body, ctx);
+                }
+                _ => continue,
+            }
+        }
+    }
+}
+
+fn find_return_statement<'a>(func_body: &OBox<'_, FunctionBody<'a>>, ctx: &LintContext<'a>) {
+    let Some(return_stmt) =
+        func_body.statements.iter().find(|stmt| matches!(stmt, Statement::ReturnStatement(_)))
```

Comment:
This will fail if a return statement is nested inside another statement, e.g.
```js
Promise.resolve(1).finally(() => {
  if (foo) {
    return true
  } else {
    return false
  }
})
```

Reply:
Feels we also need to use CFG in this rule. 

[`contains_return_statement`](https://github.com/oxc-project/oxc/blob/ea30aecb676e5cf4e1b24647d5198e36b39cc22b/crates/oxc_linter/src/rules/react/require_render_return.rs#L93-L139)

I tested `require_render_return` in our oxc-ecosystem-ci repo, didn't find falsy positive, hope it's safe to use.

Reply:
@mysteryven happy to use that function, where should I move it then to make it re-usable?

Reply:
Looks like `oxc_linter/crates/oxc_linter/src/utils/mod.rs` is good 👀

https://github.com/oxc-project/oxc/blob/c3c576660114735832fbc78e13f97a43def08593/crates/oxc_linter/src/utils/mod.rs#L36-L37